File: bloom/bloom.py
import sys
import time
from hashlib import sha256, md5, sha1
from math import log, ceil
from random import random


class BloomFilter:
    @staticmethod
    def calculate_parameters(input_size, false_positive_ratio):
        bit_length = -input_size * log(false_positive_ratio) / (log(2) ** 2)
        rounds = bit_length / input_size * log(2)
        return ceil(bit_length), ceil(rounds)

    bit_weights = [128, 64, 32, 16, 8, 4, 2, 1]

    def __init__(self, length, hash_function, rounds):
        self._bitfield = bytearray(length // 8)
        # Cannot use the 'length' parameter since it might be longer than len(self._bitfield) * 8 after division
        self._length = len(self._bitfield) * 8
        self._hash_function = hash_function
        self._rounds = rounds

    def add(self, element):
        hash_input = element
        for _ in range(self._rounds):
            hash_input = self._hash_function(hash_input)
            bit_pos = int.from_bytes(hash_input, 'big', signed=False) % self._length
            byte_pos, byte_offset = bit_pos // 8, bit_pos % 8
            self._bitfield[byte_pos] |= self.bit_weights[byte_offset]

    def is_member(self, element):
        hash_input = element
        for _ in range(self._rounds):
            hash_input = self._hash_function(hash_input)
            bit_pos = int.from_bytes(hash_input, 'big', signed=False) % self._length
            byte_pos, byte_offset = bit_pos // 8, bit_pos % 8
            if not self._bitfield[byte_pos] & self.bit_weights[byte_offset]:
                return False

        return True


def hash_func(input):
    hasher = sha1()
    hasher.update(input)
    return hasher.digest()


def demo():
    ratio = 0.01
    with open('/usr/share/dict/words') as f:
        items = f.read().split()

    start = time.time()
    items = [item.encode() for item in items]
    to_add = []
    to_check = []

    for item in items:
        if random() < ratio:
            to_check.append(item)
        else:
            to_add.append(item)

    optimal_len, optimal_rounds = BloomFilter.calculate_parameters(len(to_add), ratio)
    print(f'Item count: {len(to_add)}. Optimal bit length: {optimal_len}. Optimal rounds: {optimal_rounds}')
    bf = BloomFilter(optimal_len, hash_func, optimal_rounds)
    for item in to_add:
        bf.add(item)

    for item in to_add:
        if not bf.is_member(item):
            raise Exception

    false_positives = 0
    for item in to_check:
        if bf.is_member(item):
            false_positives += 1
    end = time.time()

    print(f'False positives: {false_positives}. False positive ratio: {float(false_positives) / len(to_check)}. Elapsed time: {end-start}')

    s = {item for item in to_add}
    d = {item: True for item in to_add}
    print(f'Total memory used by filter bitfield: {sys.getsizeof(bf._bitfield)}')
    print(f'Total memory used by set: {sys.getsizeof(s)}')
    print(f'Total memory used by dictionary: {sys.getsizeof(d)}')
    print(f'Total memory used by list: {sys.getsizeof(to_add)}')


if __name__ == '__main__':
    demo()

File: bloom/bloom.nim
import md5, os, bignum



proc hashF(input: string): string =
    $toMD5(input)


proc hashToBigInt(hash: string): bignum.Int =
    bignum.newInt(hash, 16)



type
    BloomFilter = ref object
        bitfield: seq[byte]
        hashFunc: (proc(item: string): string)
        rounds: int


proc newBloomFilter(length: int, rounds: int, hashFunc: (proc(item: string): string)): BloomFilter =
    let byteArrayLength: int = length div 8
    let byteArray = newSeq[byte](byteArrayLength)
    result = BloomFilter(
        bitfield: byteArray,
        hashFunc: hashFunc,
        rounds: rounds,
    )


proc add(bf: BloomFilter, item: string) =
    var input = item
    for _ in 1 .. bf.rounds:
        input = bf.hashFunc(input)
        let hashValue = hashToBigInt(input)
        let bitIndex = (hashValue mod (bf.bitfield.len() * 8)).toInt()
        let byteIndex = bitIndex div 8
        let byteOffset = bitIndex mod 8
        let value = byte(1 shl (7 - byteOffset))
        bf.bitfield[byteIndex] = bf.bitfield[byteIndex] or value


proc isMember(bf: BloomFilter, item: string): bool =
    var input = item
    for _ in 1 .. bf.rounds:
        input = bf.hashFunc(input)
        let hashValue = hashToBigInt(input)
        let bitIndex = (hashValue mod (bf.bitfield.len() * 8)).toInt()
        let byteIndex = bitIndex div 8
        let byteOffset = bitIndex mod 8
        let value = byte(1 shl (7 - byteOffset))
        if (bf.bitfield[byteIndex] and value) == 0:
            return false

    return true




proc getInput(filePath: string): seq[string] =
    let file = open(filePath, fmRead)
    result = @[]
    for line in file.lines:
        result.add(line)


when isMainModule:
    proc demo() =
        let words = getInput("/usr/share/dict/words")
        let bloomFilter = newBloomFilter(4000000, 7, hashF)
        var toadd = newSeq[string](0)
        var toskip = newSeq[string](0)
        for i, word in words:
            if i mod 100 == 0:
                toskip.add(word)
            else:
                toadd.add(word)

        for word in toadd:
            bloomFilter.add(word)

        var falseNegatives = 0
        for word in toadd:
            if not bloomFilter.isMember(word):
                falseNegatives += 1

        echo("False negatives: ", falseNegatives)


        var falsePositives = 0
        for word in toskip:
            if bloomFilter.isMember(word):
                falsePositives += 1

        echo("To skip: ", toskip.len())
        echo("False positives: ", falsePositives)

    demo()


